File: backend/src/models/Goal.ts
import mongoose, { Schema, Document } from 'mongoose';
import { GOAL_CATEGORIES, GOAL_PRIORITIES } from '../utils/constants';

export interface IGoal extends Document {
  userId: mongoose.Types.ObjectId;
  name: string;
  targetAmount: number;
  currentAmount: number;
  targetDate?: Date;
  icon: string;
  color: string;
  category: typeof GOAL_CATEGORIES[number];
  priority: typeof GOAL_PRIORITIES[number];
  notes?: string;
  isCompleted: boolean;
  completedAt?: Date;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const goalSchema = new Schema<IGoal>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    name: { type: String, required: true, trim: true },
    targetAmount: { type: Number, required: true },
    currentAmount: { type: Number, default: 0 },
    targetDate: Date,
    icon: { type: String, default: 'flag' },
    color: { type: String, default: '#0D6B4F' },
    category: { type: String, enum: GOAL_CATEGORIES, default: 'savings' },
    priority: { type: String, enum: GOAL_PRIORITIES, default: 'medium' },
    notes: String,
    isCompleted: { type: Boolean, default: false },
    completedAt: Date,
    isActive: { type: Boolean, default: true },
  },
  { timestamps: true },
);

goalSchema.index({ userId: 1, isActive: 1 });
goalSchema.index({ userId: 1, isCompleted: 1 });

export const Goal = mongoose.models.Goal || mongoose.model<IGoal>('Goal', goalSchema);

File: syncflow_app/lib/models/goal.dart
class Goal {
  final String id;
  final String name;
  final double targetAmount;
  final double currentAmount;
  final String? targetDate;
  final String icon;
  final String color;
  final String category;
  final String priority;
  final String? notes;
  final bool isCompleted;
  final String? completedAt;
  final bool isActive;
  final String createdAt;

  Goal({
    required this.id,
    required this.name,
    required this.targetAmount,
    this.currentAmount = 0,
    this.targetDate,
    this.icon = 'flag',
    this.color = '#0D6B4F',
    this.category = 'savings',
    this.priority = 'medium',
    this.notes,
    this.isCompleted = false,
    this.completedAt,
    this.isActive = true,
    required this.createdAt,
  });

  factory Goal.fromJson(Map<String, dynamic> json) {
    return Goal(
      id: json['_id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      targetAmount: (json['targetAmount'] as num?)?.toDouble() ?? 0,
      currentAmount: (json['currentAmount'] as num?)?.toDouble() ?? 0,
      targetDate: json['targetDate'] as String?,
      icon: json['icon'] as String? ?? 'flag',
      color: json['color'] as String? ?? '#0D6B4F',
      category: json['category'] as String? ?? 'savings',
      priority: json['priority'] as String? ?? 'medium',
      notes: json['notes'] as String?,
      isCompleted: json['isCompleted'] as bool? ?? false,
      completedAt: json['completedAt'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      createdAt: json['createdAt'] as String? ?? '',
    );
  }

  Map<String, dynamic> toCreateJson() {
    return {
      'name': name,
      'targetAmount': targetAmount,
      if (currentAmount != 0) 'currentAmount': currentAmount,
      if (targetDate != null) 'targetDate': targetDate,
      if (icon != 'flag') 'icon': icon,
      if (color != '#0D6B4F') 'color': color,
      if (category != 'savings') 'category': category,
      if (priority != 'medium') 'priority': priority,
      if (notes != null) 'notes': notes,
    };
  }

  double get percentage {
    if (targetAmount <= 0) return 0;
    return (currentAmount / targetAmount) * 100;
  }

  double get remaining => targetAmount - currentAmount;
}
